File: java/pl/edu/pg/mygame/organisms/plants/Plant.java
package pl.edu.pg.mygame.organisms.plants;

import pl.edu.pg.mygame.Grid;
import pl.edu.pg.mygame.World;
import pl.edu.pg.mygame.organisms.Organism;
import pl.edu.pg.mygame.utils.OrganismFactory;

import java.awt.Point;
import java.util.List;

import static pl.edu.pg.mygame.utils.HelperFunctions.getRandomPoint;
import static pl.edu.pg.mygame.utils.HelperFunctions.isLucky;

public abstract class Plant extends Organism {
    public Plant(Point p) {
        super(p);
    }

    @Override
    public void makeAction(Grid grid) {
        if (isLucky(3)) {
            List<Point> points = grid.getFreeNeighbourCells(position);
            if (points.size() == 0)
                return;
            Point new_point = getRandomPoint(points);
            Organism new_plant = OrganismFactory.create(species, new_point);
            grid.addOrganism(new_plant);
            World.message_panel.showMessage(species + " spread to: " +
                    String.format("(%d, %d)\n", new_point.x, new_point.y));
        }
    }

}

File: python/organisms/plants/plant.py
from abc import abstractmethod

from gui.logger import Logger
from organisms.organism import Organism
from utils.helpers import Helpers
import world


class Plant(Organism):
    @abstractmethod
    def __init__(self, pos):
        super().__init__(pos)

    def make_action(self, grid):
        if Helpers.is_lucky(3):
            points = grid.get_free_neighbor_cells(self.position)
            if len(points) == 0:
                return
            new_point = Helpers.get_random_point(points)
            world.World.create_organism(self.species, new_point, grid)
            Logger.log(self.species.name + ' spread to: ' + str(new_point))
